target_chains/ethereum/contracts/contracts/pulse/Pulse.sol

```diff
@@ -0,0 +1,462 @@
+        (bool sent, ) = msg.sender.call{value: amount}("");
+        require(sent, "withdrawal to msg.sender failed");
+
+        emit ProviderWithdrawn(provider, msg.sender, amount);
+    }
+
+    function requestPriceUpdatesWithCallback(
+        address provider,
+        uint256 publishTime,
+        bytes32[] calldata priceIds,
+        uint256 callbackGasLimit
+    ) external payable override returns (uint64 requestSequenceNumber) {
+        ProviderInfo storage providerInfo = _state.providers[provider];
+        if (providerInfo.sequenceNumber == 0) revert NoSuchProvider();
+
+        if (
+            providerInfo.maxNumPrices > 0 &&
+            priceIds.length > providerInfo.maxNumPrices
+        ) {
+            revert ExceedsMaxPrices(
+                uint32(priceIds.length),
+                providerInfo.maxNumPrices
+            );
+        }
+
+        // Assign sequence number and increment
+        requestSequenceNumber = providerInfo.sequenceNumber++;
+
+        // Verify fee payment
+        uint128 requiredFee = getFee(provider, callbackGasLimit);
+        if (msg.value < requiredFee) revert InsufficientFee();
+
+        // Store request for callback execution
+        Request storage req = allocRequest(provider, requestSequenceNumber);
+        req.provider = provider;
+        req.sequenceNumber = requestSequenceNumber;
+        req.publishTime = publishTime;
+        req.priceIds = priceIds;
+        req.callbackGasLimit = callbackGasLimit;
+        req.requester = msg.sender;
+
+        // Update fee balances
+        providerInfo.accruedFeesInWei += providerInfo.feeInWei;
+        _state.accruedPythFeesInWei +=
+            SafeCast.toUint128(msg.value) -
+            providerInfo.feeInWei;
+
+        emit PriceUpdateRequested(req);
+    }
+
+    function executeCallback(
+        address provider,
+        uint64 sequenceNumber,
+        bytes32[] calldata priceIds,
+        bytes[] calldata updateData,
+        uint256 callbackGasLimit
+    ) external payable override {
+        Request storage req = findActiveRequest(provider, sequenceNumber);
+
+        if (
+            keccak256(abi.encode(req.priceIds)) !=
+            keccak256(abi.encode(priceIds))
+        ) {
+            revert InvalidPriceIds(priceIds, req.priceIds);
+        }
+
+        if (req.callbackGasLimit != callbackGasLimit) {
+            revert InvalidCallbackGasLimit(
+                callbackGasLimit,
+                req.callbackGasLimit
+            );
+        }
+
+        PythStructs.PriceFeed[] memory priceFeeds = IPyth(_state.pyth)
+            .parsePriceFeedUpdates(
+                updateData,
+                priceIds,
+                SafeCast.toUint64(req.publishTime),
+                SafeCast.toUint64(req.publishTime)
+            );
+
+        uint256 publishTime = priceFeeds[0].price.publishTime;
+
+        try
+            IPulseConsumer(req.requester).pulseCallback(
```

Comment:
shouldn't we use the `callbackGasLimit` here?

Reply:
You need to also make sure that there's enough remaining gas to avoid a provider cancelling a request by providing a lower gas. There is a instruction called gasleft for that.